odoo/modules/packages.py

```diff
@@ -0,0 +1,281 @@
+#
+#      +---------+                                          +---------+
+#      |  base   |                                          |  base   |
+#      +---------+                                          +---------+
+#        ^   installed                                        ^    to upgrade
+#        |                                                    |
+#        |                                                    |
+#      +---------+                                          +---------+
+#      | module1 |                                          | module1 | <-----+
+#      +---------+                                          +---------+       |
+#        ^   installed                                        ^    to upgrade |
+#        |                         ==>                        |               |
+#        |                                                    |               |
+#      +---------+                                          +---------+     +---------+
+#      | module2 |                                          | module2 |     | module3 |
+#      +---------+                                          +---------+     +---------+
+#        ^   installed                                        ^    to upgrade ^    to install
+#        |                                                    |               |
+#        |                                                    |               |
+#      +---------+                                          +---------+       |
+#      | module4 |                                          | module4 | ------+
+#      +---------+                                          +---------+
+#            installed                                             to upgrade
+#
+# Because of the new dependency module4 -> module3
+# The module3 will be marked 'to install' while upgrading, and module4 should be loaded after module3
+# As a result, the updating order is
+# phase 0: base
+# phase 1: module1 -> module2
+# phase 2: module3
+# phase 3: module4
+
+
+class Package:
+    """
+    Python package for the Odoo module with the same name
+    """
+    def __init__(self, name: str, package_graph: PackageGraph) -> None:
+        # manifest data
+        self.name: str = name
+        self.manifest: dict = get_manifest(name) or {}
+
+        # ir_module_module data                     # column_name
+        self.id: int = 0                            # id
+        self.state: str = 'uninstalled'             # state
+        self.dbdemo: bool = False                   # demo
+        self.installed_version: str | None = None   # latest_version (attention: Incorrect field names !! in ir_module.py)
+
+        # info for upgrade
+        self.load_state: str = 'uninstalled'        # the state when added to package_graph
+        self.load_version: str | None = None        # the version when added to package_graph
+
+        # dependency
+        self.depends: OrderedSet[Package] = OrderedSet()
```

Comment:
but `Package` doesn't seem hashable?? How does it work?

Reply:
By object identity, which is fine here.